test_conformance/api/test_mem_object_properties_queries.cpp

```diff
@@ -0,0 +1,294 @@
+};
+
+
+int create_object_and_check_properties(cl_context context,
+                                       clMemWrapper& test_object,
+                                       test_data test_case, cl_mem_flags flags,
+                                       std::vector<cl_uint> local_data,
+                                       cl_uint size_x, cl_uint size_y)
+{
+    int error = CL_SUCCESS;
+    size_t set_size;
+    std::vector<cl_mem_properties> object_properties_check;
+    cl_image_format format;
+    format.image_channel_order = CL_RGBA;
+    format.image_channel_data_type = CL_UNSIGNED_INT32;
+    cl_image_desc desc;
+    memset(&desc, 0x0, sizeof(cl_image_desc));
+    desc.image_type = CL_MEM_OBJECT_IMAGE2D;
+    desc.image_width = size_x;
+    desc.image_height = size_y;
+
+    if (test_case.obj_t == image)
+    {
+        if (test_case.query_t == NULL_PATH)
+        {
+            test_object =
+                clCreateImageWithProperties(context, NULL, flags, &format,
+                                            &desc, local_data.data(), &error);
+        }
+        else
+        {
+            test_object = clCreateImageWithProperties(
+                context, test_case.properties.data(), flags, &format, &desc,
+                local_data.data(), &error);
+        }
+        test_error(error, "clCreateImageWithProperties failed");
+    }
+    if (test_case.obj_t == buffer)
+    {
+        if (test_case.query_t == NULL_PATH)
+        {
+            test_object = clCreateBufferWithProperties(
+                context, NULL, flags, local_data.size() * sizeof(cl_uint),
+                local_data.data(), &error);
+        }
+        else
+        {
+            test_object = clCreateBufferWithProperties(
+                context, test_case.properties.data(), flags,
+                local_data.size() * sizeof(cl_uint), local_data.data(), &error);
+        }
+
+        test_error(error, "clCreateBufferWithProperties failed.");
+    }
+    clGetMemObjectInfo(test_object, CL_MEM_PROPERTIES, 0, NULL, &set_size);
+    test_error(error,
+               "clGetMemObjectInfo failed asking for CL_MEM_PROPERTIES.");
+    if (test_case.query_t == NULL_PATH && set_size == 0)
+    {
+        return TEST_PASS;
+    }
+    if (test_case.query_t == NON_NULL_PATH && set_size == 0)
+    {
+        log_error("clGetMemObjectInfo - invalid value returned");
+        return TEST_FAIL;
+    }
+    cl_uint number_of_props = set_size / sizeof(cl_mem_properties);
+    object_properties_check.resize(number_of_props);
+    clGetMemObjectInfo(test_object, CL_MEM_PROPERTIES, set_size,
+                       object_properties_check.data(), NULL);
+    test_error(error,
+               "clGetMemObjectInfo failed asking for CL_MEM_PROPERTIES.");
+
+    if (object_properties_check != test_case.properties)
+    {
+        log_error("clGetMemObjectInfo returned different properties than was "
+                  "set initially.\n");
+        error = TEST_FAIL;
+    }
```

Comment:
The behavior being tested here is valid behavior but I don't think it's the _only_ valid behavior, see https://github.com/KhronosGroup/OpenCL-Docs/issues/317.

Since we haven't defined any new properties in OpenCL 3.0, I think we can only test that the query executed successfully:

* It's valid to return "0" for the size of the queried properties, since we have not defined any properties.  This is valid regardless whether the passed-in properties are `NULL` or an empty property list consisting only of the "0" terminator.
* It's valid to return a single property value of "0", indicating an empty property list.  Again, this is valid regardless whether the passed-in properties are `NULL` or an empty property list consisting only of the "0" terminator.
* It's valid to return additional default properties, say if an extension adds new properties with default values.

I suppose we could test that the property list consists of property pairs and is terminated by zero?  We can't check for equality with the passed-in properties, though.

Reply:
This thread has been superseded by my comment below (https://github.com/KhronosGroup/OpenCL-CTS/pull/857#discussion_r463286622).